components/fonts/platform/freetype/ohos/font_list.rs

```diff
@@ -68,23 +105,265 @@ struct FontList {
+    } else {
+        None
+    }
+}
+
+// Note: The weights here are taken from the `alias` section of the fontconfig.json
+fn detect_hos_font_weight_alias(font_modifiers: &[&str]) -> Option<i32> {
+    if font_modifiers.contains(&"Light") {
+        Some(100)
+    } else if font_modifiers.contains(&"Regular") {
+        Some(400)
+    } else if font_modifiers.contains(&"Medium") {
+        Some(700)
+    } else if font_modifiers.contains(&"Bold") {
+        Some(900)
+    } else {
+        None
+    }
+}
+
+fn noto_weight_alias(alias: &str) -> Option<i32> {
+    match alias.to_ascii_lowercase().as_str() {
+        "thin" => Some(100),
+        "extralight" => Some(200),
+        "light" => Some(300),
+        "regular" => Some(400),
+        "medium" => Some(500),
+        "semibold" => Some(600),
+        "bold" => Some(700),
+        "extrabold" => Some(800),
+        "black" => Some(900),
+        _unknown_alias => {
+            warn!("Unknown weight alias `{alias}` encountered.");
+            None
+        },
+    }
+}
+
+fn detect_hos_font_width(font_modifiers: &[&str]) -> FontWidth {
+    if font_modifiers.contains(&"Condensed") {
+        FontWidth::Condensed
+    } else {
+        FontWidth::Normal
+    }
+}
+
+/// Split a Noto font filename into the family name with spaces
+///
+/// E.g. `NotoSansTeluguUI` -> `Noto Sans Telugu UI`
+fn split_noto_font_name(name: &str) -> Vec<String> {
+    let mut name_components = vec![];
+    let mut current_word = String::new();
+    let mut chars = name.chars();
+    // To not split acronyms like `UI` or `CJK`, we only start a new word if the previous
+    // char was not uppercase.
+    let mut previous_char_was_uppercase = true;
+    if let Some(first) = chars.next() {
+        current_word.push(first);
+        for c in chars {
+            if c.is_uppercase() {
+                if !previous_char_was_uppercase {
+                    name_components.push(current_word.clone());
+                    current_word = String::new();
+                }
+                previous_char_was_uppercase = true;
+            } else {
+                previous_char_was_uppercase = false
+            }
+            debug_assert!(
+                !c.is_whitespace(),
+                "split_noto_font_name() assumes the font filename does not contain whitespace"
+            );
+            current_word.push(c)
+        }
+    }
+    if current_word.len() > 0 {
+        name_components.push(current_word);
+    }
+    name_components
+}
+
+/// Parse the font file names to determine the available FontFamilies
+///
+/// In principle the `fontconfig.json` should provide information on the available fonts,
+/// but on some devices the `font_file_map` member is missing, and in general the family name to
+/// file name mapping seems to not be accurate, so we instead infer the family name from the
+/// file names of the installed fonts.
+fn parse_font_filenames(font_files: Vec<PathBuf>) -> Vec<FontFamily> {
```

Comment:
There's a lot of code here to parse font names. How many fonts does this deal with in practice? Perhaps it would be simpler to simply have a pre-calculated list of fonts and with their weights? This parsing isn't always going to work in practice. For instance `NotoSerif[wdth,wght].ttf` likely contains multiple weights (perhaps named variants of using OpenType variations -- you can determine this with FreeType. Parsing the name is never going to give you access to this.

Reply:
> How many fonts does this deal with in practice? 

I recently updated my phone to a newer pre-release version and now there are 193 fonts in `/system/fonts`, 180 of which are Noto fonts. Previously there were a lot less Noto fonts.
The filenames of the HarmonyOS fonts also changed compared to the last version, and the available OpenHarmony versions - the previously present "Regular" suffix is now omitted in the filenames of HarmonyOS Sans fonts.

> For instance NotoSerif[wdth,wght].ttf likely contains multiple weights (perhaps named variants of using OpenType variations -- you can determine this with FreeType

How much support is there currently in servo for variable font features on other platforms? I guess I would need to manually construct the `FontTemplateDescriptor`, since the constructor assumes non-variable fonts.
Is correctly setting the weight range mandatory, or could it also be left at the default value (for now), and we would just lose the variable aspect of the fonts (but without anything breaking in wierd ways)?

The main reason I went with parsing the font names here, is that I somehow need to get the family names. On the most recent version I'm running there seems to be a newly added `font_file_map` map in the `fontconfig.json` file, which might be an option if we require OpenHarmony 5.0 and newer. But at least in my current preview build that config file is still  a bit buggy, and many font filenames in the map don't exist on the actual device.

Reply:
@mrobinson For now I at least hardcoded the paths for the HMOS color emojis, since their paths haven't changed. However, I would still like to keep the parsing logic to support the current OH 5.0 beta, at least until an update has been released, which provides the correct filenames in the `fontconfig.json` (It's already been fixed, should be in the next beta version, whenever that is released)

Reply:
> How much support is there currently in servo for variable font features on other platforms? I guess I would need to manually construct the `FontTemplateDescriptor`, since the constructor assumes non-variable fonts.
> Is correctly setting the weight range mandatory, or could it also be left at the default value (for now), and we would just lose the variable aspect of the fonts (but without anything breaking in wierd ways)?

Apologies for the late reply: Servo has supported for named variations (which are usually stored in font tables), but not support for OpenType variations. Named variations are usually expressed as the font index. These are the kind that are typically returned via FreeType/FontConfig.